test/hotspot/jtreg/compiler/c2/irTests/XorLNodeIdealizationTests.java

```diff
@@ -217,4 +229,57 @@ public long test16(long x, long y, long z) {
     public long test17(long x) {
         return ~(x + 2022);
     }
+
+    @Test
+    @IR(failOn = {IRNode.XOR})
+    @IR(counts = {IRNode.CON_L, "1"})
+    // Checks (c ^c)  => c (constant folded)
+    public long testConstXor() {
+        return CONST_1 ^ CONST_2;
+    }
+
+    @Test
+    @IR(failOn = {IRNode.XOR})
+    @IR(counts = {IRNode.CON_L, "1"})
+    // Checks (x ^ x)  => c (constant folded)
+    public long testXorSelf(long x) {
+        return x ^ x;
+    }
+
+    // clamp value to [1,CONST_POW_2]
+    @ForceInline
+    private static long forceMinMax(long value) {
+        // equivalent to Math.min(CONST_POW_2, Math.max(value, 1))
+        return 1L + (value & (CONST_POW_2 - 1L));
+    }
+
+    @Test
+    @IR(counts = {IRNode.XOR, "1"})  // must not be constant-folded
+    // checks that add_ring computes correct max on exact powers of 2
+    public boolean testMaxPow2(long x, long y) {
+        x = forceMinMax(x);
+        y = forceMinMax(y);
+        long xor = x ^ y;
+        return xor < CONST_POW_2;
+    }
+
+    @DontCompile
+    public boolean interpretedMaxPow2(long x, long y) {
+        x = forceMinMax(x);
+        y = forceMinMax(y);
+
+        long xor = x ^ y;
+        return xor < CONST_POW_2;
+    }
+
+    @Test
+    @IR(failOn = {IRNode.XOR})
+    @IR(counts = {IRNode.CON_I, "1"})
+    public boolean testMaxPow2Folded(long x, long y) {
+        x = forceMinMax(x);
+        y = forceMinMax(y);
+
+        long xor = x ^ y;
+        return xor < (CONST_POW_2*2L);
+    }
```

Comment:
Thanks for adding range-tests like these!
I think we need some smarter tests on top of these though.

For one, we should have some camped ranges that are not restricted to `[1,CONST_POW_2]`, but have any lower and upper limit.

Further: we should check with `if` branches if the resulting range is correct. Not that we get the resulting range off-by-1 and have some edge case that constant-folds the wrong way ;)